lucene/core/src/java/org/apache/lucene/search/ScorerUtil.java

```diff
@@ -116,6 +116,25 @@ public int length() {
     }
   }
 
+  /**
+   * Compute a minimum required score, so that (float) MathUtil.sumUpperBound(minRequiredScore +
+   * maxRemainingScore, numScorers) <= minCompetitiveScore. The computed value may not be the
+   * greatest value that meets this condition, which means that we may fail to filter out some docs.
+   * However, this doesn't hurt correctness, it just means that these docs will be filtered out
+   * later, and the extra work required to compute an optimal value would unlikely result in a
+   * speedup.
+   */
+  static double minRequiredScore(
+      double maxRemainingScore, float minCompetitiveScore, int numScorers) {
+    double minRequiredScore = minCompetitiveScore - maxRemainingScore;
+    double subtraction = Math.ulp((double) minCompetitiveScore);
```

Comment:
```suggestion
    double subtraction = Math.ulp(minCompetitiveScore); // note: we want the float ulp, not the double ulp
```

Sorry, I realize that the current code was my suggestion, but looking at it again, we need to take the float ulp to make sure it converges quickly.